src/ir/var.rs

```diff
@@ -130,6 +130,53 @@ fn default_macro_constant_type(value: i64) -> IntKind {
     }
 }
 
+/// Determines whether a set of tokens from a CXCursor_MacroDefinition
+/// represent a function-like macro. If so, calls the func_macro callback
+/// and returns `Err(ParseError::Continue)` to signal to skip further
+/// processing.  If conversion to UTF-8 fails (it is performed only where it
+/// should be infallible), then `Err(ParseError::Continue)` is returned as well.
+fn handle_function_macro(
+    tokens: &[clang::ClangToken],
+    callbacks: &dyn crate::callbacks::ParseCallbacks,
+) -> Result<(), ParseError> {
+    let mut spelled = tokens.into_iter().map(clang::ClangToken::spelling);
+    let boundary = spelled
+        .clone() // clone the iterator to avoid consuming elements
+        .position(|s| s == b")");
+
+    // cexpr explicitly does not handle function-like macros, except for a
+    // degenerate case in which it behaves in a non-conformant way. We prevent
+    // cexpr from ever seeing function-like macros by checking for a parenthesis
+    // token immediately adjacent to (that is, abutting) the first token in the
+    // macro definition.
+    match (tokens.get(0..2), boundary) {
```

Comment:
nit: I'd prefer for this to early-return instead of nesting a giant `match`, I think it's nicer.

So:

```rust
let boundary = match tokens.iter().position(|token| token.is_closing_paren()) {
    Some(b) => b,
    None => return Ok(()),
};

let (first, second) = match tokens.get(0..2) {
    Some([first, second]) => (first, second),
    _ => return Ok(()),
};

if !first.is_abutting(second) || !second.is_opening_paren() {
    return Ok(());
}

// At this point we know we have a function macro, do all the expensive stuff as needed.
```

Reply:
I agree, your formulation is much nicer to read.

By chance, I was already refactoring again while you were reviewing, so by the time I read this thread, I had come up with something in 2ffe97bebf46192f74c1ff4764085e1d7dbd65b9 that does an early return, but differently.

(I found out about [`clang_Cursor_isMacroFunctionLike`](https://clang.llvm.org/doxygen/group__CINDEX__TYPES.html#ga0fcb19b77fa3eb1dd531bb8f20f65e6c) from https://github.com/jethrogb/rust-cexpr/issues/3#issuecomment-308905175 and https://github.com/rust-lang/rust-bindgen/issues/753#issuecomment-308901159, so I think my new code avoids checking the first two tokens at all, except on libclang < 3.9.)

I will leave this item open until you say whether my new formulation still needs help.

Reply:
I think your subsequent statements elsewhere mean I can resolve this, so that is what I am doing.